# DP: Fix PR c/33763, just disabling the sorry message.

--- a/src/gcc/tree-inline.c
+++ b/src/gcc/tree-inline.c
@@ -3836,6 +3836,12 @@
 	goto egress;
 
       if (lookup_attribute ("always_inline", DECL_ATTRIBUTES (fn))
+	  /* For extern inline functions that get redefined we always
+          silently ignored alway_inline flag. Better behaviour would
+          be to be able to keep both bodies and use extern inline body
+          for inlining, but we can't do that because frontends overwrite
+          the body.  */
+	  && !cg_edge->callee->local.redefined_extern_inline
 	  /* Avoid warnings during early inline pass. */
 	  && cgraph_global_info_ready
 	  /* PR 20090218-1_0.c. Body can be provided by another module. */
